kernel/vm.c

```diff
@@ -102,6 +102,27 @@ walk(pagetable_t pagetable, uint64 va, int alloc)
   return &pagetable[PX(0, va)];
 }
 
+void printwalk(pagetable_t pagetable, uint64 level)
+{
+  // there are 2^9 = 512 PTEs in a page table.
+  for(int i = 0; i < 512; i++){
+    pte_t pte = pagetable[i];
+    
+    if (pte & PTE_V)
+    {
+      for (int j = 0; j < level; ++j) printf("... ");
+      printf("   ");
+      printf("%d: pte %p pa %p %s%s%s  ||  %s\n", i, pte, PTE2PA(pte), (pte) & PTE_R ? "r" : "-", (pte) & PTE_W ? "w" : "-", (pte) & PTE_X ? "x" : "-", (char*)(PTE2PA(pte)));
```

Comment:
Very long line, so hard to understand what is going on. Consider to extract values to named variables or change formatting.